### PR TITLE
cyberuskey lib codebase

### PR DESCRIPTION
**Pull request by bkozub**

### CODE REVIEW

**Comment by bkozub** [comment]
**TODO**:
- [x] Prepare as a package, add setup.py
- [x] Update README.MD with usage example
- [x] Update required package list
- [x] Add some comments and function description

File: cyberuskey/cyberuskey.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+class CyberusKey:
+    client_id: str
+    client_secret: str
+    redirect_uri: str
+    _API_URI: str = "https://api.cyberuskey.com"
+    _OPENID_PUBLIC: str = "-----BEGIN PUBLIC KEY-----\nMIGeMA0GCSqGSIb3DQEBAQUAA4GMADCBiAKBgHElKnuERpCN/WcD6RtS9rKhJODM\nIdr2Y1yFrS255cOaG10CLwFPhSVK5z4HQv5/VN3GB2Ft+fbu9OZRTqdA4lHo0PB3\nKaj3yByDUdIoTHd4RmZMLSFVHKR0KAW193nI7s/pzeqDL0oFpHnRNZGUqhRbm2UK\nfHHDWKkTn/iGIV7XAgMBAAE=\n-----END PUBLIC KEY-----"
+    __access_token: str = field(init=False, default=None)
+    __id_token: str = field(init=False, default=None)
+
+    def _is_uri(self, uri: str):
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: let's keep private/protected methods at the bottom of the classes

File: cyberuskey/cyberuskey.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    def _validate_claim_hash(self, original_hash: str, value: str) -> bool:
+        computed_hash = self._compute_claim_hash(value)
+
+        if original_hash != computed_hash:
+            return False
+
+        return True
+
+    def authenticate(
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: I know, I know, it's misleading, but let's rename it to **authorize**. In OIDC spec you have Authorization Endpoint (https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationEndpoint). In fact, that's true, we identify the user and her/his role (like client admin). In the context of the client, we identify it and check whether it has access to some resources (like redirect uri).

File: cyberuskey/cyberuskey.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        self,
+        query_arguments: dict = None,
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: Use Dict from typings

Comment (arturwojnar-cyberus) [reply cyberuskey/cyberuskey.py]: Let's add pylint here and some formatter.

File: cyberuskey/cyberuskey.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        state: str = None,
+        nonce: str = None,
+        code: str = None,
+        error: str = None,
+        error_description: str = None,
+    ) -> dict:
+        #TODO some description
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: Let's do it

File: cyberuskey/cyberuskey.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            raise AuthenticateException(error, error_description)
+
+        code = query_arguments.get("code") or code
+        if not code:
+            raise MissingAuthorizationCode
+
+        if isinstance(code, list):
+            code = code[0].decode("utf-8")
+
+        if state:
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: state - get it from query_arguments, the same way as code/error

File: cyberuskey/cyberuskey.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        error_description: str = None,
+    ) -> dict:
+        #TODO some description
+
+        if query_arguments is None:
+            query_arguments = {}
+
+        error = query_arguments.get("error") or error
+        if error:
+            raise AuthenticateException(error, error_description)
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py:111]: error_description -> you miss taking it from the query

File: cyberuskey/cyberuskey.py (hunk 3 above)
Comment (arturwojnar-cyberus) [reply cyberuskey/cyberuskey.py]: Hm, you do have no dependencies defined

File: cyberuskey/cyberuskey.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+        if token_body.get('error'):
+            return token_body['error']
+
+        id_token = token_body['id_token']
+        self.__id_token = id_token
+        id_data = jwt.decode(id_token, self._OPENID_PUBLIC, algorithms=['RS256'], audience=self.client_id)
+
+        if id_data.get("nonce") and nonce:
+            if id_data['nonce'] != nonce:
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: inconsistency, ' ' vs " "

File: cyberuskey/cyberuskey.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            if not self._validate_claim_hash(original_at_hash, access_token):
+                raise AuthenticateException("invalid_at_hash", "Access token hash is invalid")
+
+        if original_c_hash:
+            if not self._validate_claim_hash(original_c_hash, code):
+                raise AuthenticateException("invalid_c_hash", "Authorization code hash is invalid")
+
+        self.__access_token = access_token
+
+        return id_data
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: I'd say you can just return both id_data, access_token (tuple)

File: cyberuskey/cyberuskey.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        return self.__access_token
+
+    @property
+    def id_token(self) -> str:
+        if self.__id_token is None:
+            raise InvalidAuthenticateValueError
+
+        return self.__id_token
+
+    def _compute_claim_hash(self, value: str) -> str:
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: This you can have in separate util

File: cyberuskey/cyberuskey.py (hunk 3 above)
Comment (CyberusLabs) [reply cyberuskey/cyberuskey.py]: It's formatted by black

File: cyberuskey/cyberuskey.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        :return Tuple (Dict: {
+            str: 'iss': client domain,
+            str: 'sub': user openid identifier,
+            str: 'aud': client key,
+            str: 'alg': jwt decode algorithm, in our case 'RS256',
+            int: 'exp': token expiration timestamp,
+            int: 'iat': token generation timestamp,
+            str: 'mobile': app instance identifier,
+            str: 'at_hash': access token claim hash, base64 encoded from first 128 bytes,
+            str: 'c_hash': authorization code claim hash, the same as below,
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: above :D

File: cyberuskey/cyberuskey.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            raise AuthenticateException(error, error_description)
+
+        code = query_arguments.get("code") or code
+        if not code:
+            raise MissingAuthorizationCode
+
+        if isinstance(code, list):
+            code = code[0].decode("utf-8")
+
+        if state:
```
Comment (arturwojnar-cyberus) [inline cyberuskey/cyberuskey.py]: Bartek - take state from the query and nonce

**Pull request merged**